File: content/courses/parallel-computing-introduction/codes/mpi_io_fileview.f90
program mpiwrite
   use mpi_f08
   implicit none

   integer            :: i,j
   character(len=80)  :: arg
   integer            :: numargs

   integer            :: nrows, ncols
   integer            :: nrl, ncl, lrow, lcol
   integer            :: N, M

   integer            :: rank, nprocs
   integer            :: mpi_err
   integer, parameter :: root=0
   type(MPI_Status)   :: mpi_stat
   type(MPI_File)     :: fh
   type(MPI_Datatype) :: locarray
   integer            :: amode
   integer            :: lusize
   integer, dimension(2) :: gdims, ldims, starts
   integer(kind=MPI_OFFSET_KIND) :: offset=0
   integer, allocatable, dimension(:,:) :: loc_u, u, rbuf
   character(len=24)  :: fname
   character(len=80)  :: myfile

   ! check number of parameters and read in filename
   ! all ranks do this, avoids broadcast
   numargs=command_argument_count()
   if (numargs .lt. 1) then
      stop 'USAGE: output-file'
   else
      call get_command_argument(1,fname)
   endif

   !Initialize MPI
   call MPI_INIT()
   call MPI_COMM_SIZE(MPI_COMM_WORLD,nprocs)
   call MPI_COMM_RANK(MPI_COMM_WORLD,rank)

   ! Hard-code sizes so we can see what we're doing
   ! Row/column layout for ranks
   nrows=4
   ncols=4

   nrl = 4
   ncl = 4

   N=nrl*nrows
   M=nrl*ncols

   !Set up the topology
   lrow=rank/ncols
   lcol=mod(rank,ncols)

   allocate(loc_u(nrl,ncl))
   loc_u=(rank+1)

   !Write each segment to conventional file (the "old way")
   write(myfile,'(a,i2.2)') fname(1:len_trim(fname)),rank

   open(unit=10,file=myfile,status='unknown')
   do i=1,nrl
      write (10,'(*(g0,1x))') loc_u(i,:)
   enddo
   close(10)

   gdims=[N,M]
   ldims=[nrl,ncl]
   starts=[ncl*lrow,nrl*lcol]
   write(*,*) rank,starts

   call MPI_TYPE_CREATE_SUBARRAY(2, gdims, ldims, starts, MPI_ORDER_FORTRAN,   &
                                                          MPI_INTEGER, locarray)

   call MPI_TYPE_COMMIT(locarray)

   amode=ior(MPI_MODE_CREATE, MPI_MODE_WRONLY)
   call MPI_FILE_OPEN(MPI_COMM_WORLD,trim(fname),amode,MPI_INFO_NULL,fh,mpi_err)

   if ( mpi_err /= MPI_SUCCESS) then
       call MPI_FINALIZE()
       stop "Unable to open MPI file for writing"
   endif

   offset=0
   call MPI_FILE_SET_VIEW(fh, offset, MPI_INTEGER, locarray,                   &
                                    'native', MPI_INFO_NULL)

   call MPI_FILE_WRITE_ALL(fh,loc_u, size(loc_u), MPI_INTEGER, mpi_stat)
   call MPI_FILE_CLOSE(fh)

   !All processes read file
   amode=MPI_MODE_RDONLY
   call MPI_FILE_OPEN(MPI_COMM_WORLD,trim(fname),amode,MPI_INFO_NULL,fh,mpi_err)
   if ( mpi_err /= MPI_SUCCESS) then
       call MPI_FINALIZE()
       stop "Unable to open MPI file for reading"
   endif

   allocate(u(nrl,ncl),rbuf(nrl,ncl))
   rbuf=0
   lusize=size(u)

   call MPI_FILE_SET_VIEW(fh, offset, MPI_INTEGER, locarray,                   &
                                    'native', MPI_INFO_NULL)

   call MPI_FILE_READ_ALL(fh, rbuf, lusize, MPI_INTEGER, mpi_stat)
   call MPI_FILE_CLOSE(fh)

  !Usual trick to print one rank at a time
   if (rank==0) then
       write(*,*) 'Result for rank 0'
       do j=1,nrl
          write(*,'(*(i10))') rbuf(j,:)
       enddo
       do i=1,nprocs-1
          u=0
          call MPI_Recv(u,lusize, MPI_INTEGER,MPI_ANY_SOURCE,                  &
                                 MPI_ANY_TAG, MPI_COMM_WORLD,mpi_stat)
          write(*,*) 'Result for rank',mpi_stat%MPI_SOURCE
          do j=1,nrl
             write(*,'(*(i10))') u(j,:)
          enddo
       enddo
   else
       call MPI_Send(rbuf,lusize,MPI_INTEGER,0,rank,MPI_COMM_WORLD)
   endif

   call MPI_TYPE_FREE(locarray)

   call MPI_Finalize()

end program

File: content/courses/parallel-computing-introduction/codes/mpi_io_fileview.cxx
#include <cstring>
#include <cstdio>
#include <iostream>
#include <fstream>
#include <string>
#include <mpi.h>

using namespace std;

int main (int argc, char *argv[]) {

    // Declarations for MPI
    int rank, nprocs;
    int errcode;
    MPI_Status mpi_stat;
    MPI_Info info;
    MPI_Offset offset;
    MPI_File fh;
    int root=0, tag=0;
    int mpi_err;

    // Check number of parameters and read in filename
    if (argc < 2) {
       printf ("USAGE:  %s output-file\n", argv[0]);
    exit(1);
    }
    const char *fname=argv[1];

    //Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    // Hard-code sizes so we can see what we're doing
    // Row/column layout for ranks
    int nrows=4;
    int ncols=4;

    int nrl = 4;
    int ncl = 4;

    int N=nrl*nrows;
    int M=nrl*ncols;

    //Set up the topology
    int lrow=rank/ncols;
    int lcol=rank%ncols;

    int **loc_u=new int*[nrl];
    int *luptr=new int[(nrl)*(ncl)];

    for (int i=0;i<nrl;++i,luptr+=ncl) {
       loc_u[i] = luptr;
    }

    for ( int i = 0; i < nrl; i++ ) {
        for (int j = 0; j < ncl; j++ ) {
             loc_u[i][j] = rank+1;
         }
    }

    int lusize=nrl*ncl;

    //Write each segment to conventional file (the "old way")
    ofstream fp;
    string myfile=argv[1]+to_string(rank);
    fp.open(myfile,ios::out);
    for (int i = 0; i < nrl; i++) {
        for (int j = 0; j < ncl; j++) {
            fp<<loc_u[i][j]<<" ";
        }
        fp<<"\n";
    }
    fp.close();

    int gdims[]={N,M};
    int ldims[]={nrl,ncl};
    int starts[]={ncl*lrow,nrl*lcol};
    cout<<"Starts for rank "<<rank<<" are  "<<starts[0]<<" "<<starts[1]<<endl;

    MPI_Datatype locarray;
    MPI_Type_create_subarray(2,gdims,ldims,starts,MPI_ORDER_C,MPI_INT,&locarray);
    MPI_Type_commit(&locarray);

    int amode=MPI_MODE_CREATE | MPI_MODE_WRONLY;
    mpi_err=MPI_File_open(MPI_COMM_WORLD,fname,amode,MPI_INFO_NULL,&fh);

    if (mpi_err != MPI_SUCCESS) {
        cout<<"Unable to open file for writing\n";
        MPI_Finalize();
        exit(2);
    }

    offset=0;
    MPI_File_set_view(fh, offset, MPI_INT, locarray, "native", MPI_INFO_NULL);

    MPI_File_write_all(fh,&loc_u[0][0], lusize , MPI_INT, &mpi_stat);
    MPI_File_close(&fh);

    //All processes read file
    amode=MPI_MODE_RDONLY;
    mpi_err=MPI_File_open(MPI_COMM_WORLD,fname,amode,MPI_INFO_NULL,&fh);
        if ( mpi_err != MPI_SUCCESS) {
       MPI_Finalize();
       cout<<"Unable to open MPI file for reading\n";
       exit(2);
    }

    int **rbuf=new int*[nrl];
    int *rptr=new int[(nrl)*(ncl)];

    for (int i=0;i<nrl;++i,rptr+=ncl) {
       rbuf[i] = rptr;
    }

    for ( int i = 0; i < nrl; i++ ) {
        for (int j = 0; j < ncl; j++ ) {
             rbuf[i][j] = 0;
         }
    }

    MPI_File_set_view(fh, offset, MPI_INT, locarray, "native", MPI_INFO_NULL);
    MPI_File_read_all(fh, &rbuf[0][0], lusize, MPI_INT, &mpi_stat);
    MPI_File_close(&fh);

    //Usual trick to print one rank at a time
    int position;
    int *u =  new int[lusize];

    if ( rank == 0 ) {
        cout<<"Result for rank 0\n";
        for (int i=0;i<nrl;i++) {
            for (int j=0;j<ncl;j++) {
                cout<<rbuf[i][j]<<" ";
            }
            cout<<endl;
        }
        for (int n=1;n<nprocs;++n) {
           memset(u,0.,lusize);
           MPI_Recv(u,lusize,MPI_INT,MPI_ANY_SOURCE,MPI_ANY_TAG,MPI_COMM_WORLD,&mpi_stat);
           cout<<"Result for rank "<<mpi_stat.MPI_SOURCE<<endl;
           position=0;
           for (int i=0;i<nrl;i++) {
               for (int j=0;j<ncl;j++) {
                   cout<<u[position++]<<" ";
               }
               cout<<endl;
           }
       }
    }
    else {
        // Pack the 2D array into the buffer
        position=0;
        for (int i=0; i<nrl; i++)
            for (int j=0;j<ncl; j++)
                u[position++]=rbuf[i][j];

        MPI_Send(u,lusize, MPI_INT, 0, rank, MPI_COMM_WORLD);
    }

    MPI_Type_free(&locarray);

    MPI_Finalize();

}
